Tahi-project/tahi
# APERTA-5372 attach files to ae invite

JIRA issue: https://developer.plos.org/jira/browse/APERTA-5372
#### What this PR does:

This PR does a few major things.
- Invitations for Academic Editors and Reviewers are now created in a 'Draft' state on the client.  Client tasks can opt into this behavior by saving a new invitation with `state: 'pending'`.  Invitations have a `.send()` method on them that calls a new custom endpoint.
- The UI for inviting Academic Editors and Reviewers has been changed substantially.  **Note that for the purposes of this PR the gritty details of the UI are out of scope** .  They'll be polished up as part of https://developer.plos.org/jira/browse/APERTA-5588.
- Users can attach files to Academic Editor invitations while the invite is in a draft state.  The mechanics for attaching files to invitations are very close what we currently do with Adhoc Attachments.
#### Notes

We decided to forgo spending further time to unify the attachment models on the front end.  Even though the rails models are polymorphic (attachments belong to an owner rather than a task, etc) the frontend expresses the relationship between the different attachment subclasses in a few different ways.  Ideally we'd mimic the same structure as the back end but it was prohibitive for us time-wise.  We also felt like we'd benefit from taking a step back before returning to it.
#### Major UI changes

The UIs for both Academic Editor and Reviewer invites have been revamped and unified.  Many places that used a `.invitation` class will now refer to a `.invitation-item` class.  Classes that were specific to reviewers or editors have mostly been removed.

For this PR the compose screen is largely unchanged. It'll be brought in line with the mocks as part of APERTA-5588
![image](https://cloud.githubusercontent.com/assets/2043348/17864427/116c0dcc-686c-11e6-90f8-41c4357503d8.png)

Users can edit unsent invitations:
![image](https://cloud.githubusercontent.com/assets/2043348/17864471/3b5ef50e-686c-11e6-841f-5786dd3dfec3.png)

Sent emails have files attached directly:
![image](https://cloud.githubusercontent.com/assets/2043348/17864557/8853fbca-686c-11e6-8142-98899bf70717.png)

Users can still view files on sent emails:
![image](https://cloud.githubusercontent.com/assets/2043348/17864572/97be3148-686c-11e6-941d-31c6ebf31710.png)
#### How to test this:
- Open the Paper Editors card
- Compose an invite to a valid looking email address.
- You'll be able to attach files to the email in the compose screen if you'd like.
- You can hit 'close' to keep the email as a draft, or 'send' to send it immediately (that part is kind of out of scope)
- For an unsent invitation you can click on the invitation and edit the body.  You can add or remove files at this point.
- For a sent invitation you should only be able to view the list of attachments, not modify them.

---
#### Code Review Tasks:

Author tasks:
- [ ] ~~If a data-migration rake task is needed, the task is found in `lib/tasks/data-migrations` within the `data:migrate` namespace. Example task name: `aperta_9999_migration_description`~~ no existing data to migrate

Reviewer tasks:
#### After the Code Review:

Author tasks:
